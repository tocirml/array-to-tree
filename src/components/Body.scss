.App-body {
  padding: 5px;
  .body-message {
    margin: 10px;
  }
  .OutputContainer {
    padding: 10px;
    border: 1px solid black;
    background: rgb(70, 69, 69);
    overflow-x: scroll;
    left: 3px;
  }
}
