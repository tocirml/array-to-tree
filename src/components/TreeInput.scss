.tree-source {
  padding: 5px;
  .array-input {
    width: 100%;
    max-width: 400px;
    height: 30px;
  }
}
