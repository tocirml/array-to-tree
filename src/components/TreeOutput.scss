.treeNode {
  // fill me in
  padding: 5px;
  border: solid 1px rgb(0, 16, 247);
  width: auto;
  .nodeId {
    font-weight: bold;
    color: rgb(255, 255, 255);
    padding: 5px;
    text-align: center;
  }

  .nodeChildren {
    border: solid 1px rgb(251, 255, 0);
    padding: 5px;
    display: flex;
    justify-content: space-around;
    .treeNode {
      width: auto;
    }
  }
}

.smallest-subtree {
  border: 2px solid green;
}
