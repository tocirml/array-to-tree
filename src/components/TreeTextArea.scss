.tree-text-area {
  margin-top: 20px;
}
