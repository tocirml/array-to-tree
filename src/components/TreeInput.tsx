import * as React from 'react';
import { BinTreeNode } from '../models/TreeNode';
import './TreeInput.scss';

export interface TreeInputProps {
  onChange: (newTreeNode: BinTreeNode) => void;
}
interface TreeInputState {
  treeText: string | ArrayBuffer | null;
}

export class TreeInput extends React.Component<TreeInputProps, TreeInputState> {
  constructor(props: TreeInputProps) {
    super(props);
    this.state = {
      treeText: '',
    };
  }

  /**
   * Converts array format binary tree notation to BinTreeNode data structure
   * @param arrayFormat [id, leftChild, rightChild] for example [1, [2], [3, null, [5]]]
   * @returns TreeNode format
   * */
  parseArrayToTree(arrayFormat: any[]): BinTreeNode {
    // recursive solution to convert from array to json tree BinTreeNode
    if (arrayFormat.length === 1) return { id: arrayFormat[0] };

    return {
      id: arrayFormat[0],
      left: arrayFormat[1] ? this.parseArrayToTree(arrayFormat[1]) : null,
      right: arrayFormat[2] ? this.parseArrayToTree(arrayFormat[2]) : null,
    };
  }

  convert = () => {
    // After you implement parseArrayToTree above, uncomment the below code
    let treeArrayFormat: any[];
    if (typeof this.state.treeText === 'string') {
      treeArrayFormat = JSON.parse(this.state.treeText);
      this.props.onChange(this.parseArrayToTree(treeArrayFormat));
    }

    // After you implement parseArrayToTree above, comment the below code
    // let treeNodeFormat: BinTreeNode = JSON.parse(this.state.treeText);
    // this.props.onChange(treeNodeFormat);
  };

  handleChange = (e: any) => {
    const fileReader = new FileReader();
    fileReader.readAsText(e.target.files[0], 'UTF-8');
    fileReader.onload = e => {
      this.setState({
        treeText: e.target!.result,
      });
    };
  };

  render() {
    return (
      <div className="tree-source">
        <label htmlFor="array-input">Tree source</label>
        <br />
        <input
          type="text"
          name="array-input"
          className="array-input"
          value={String(this.state.treeText)}
          onChange={ev => {
            this.setState({
              treeText: ev.target.value,
            });
          }}
        />
        {/* <div className="or">OR</div> */}
        <br />
        <input type="file" onChange={this.handleChange} />
        <br />
        <button onClick={this.convert} className="fetch-button">
          Fetch
        </button>
      </div>
    );
  }
}
