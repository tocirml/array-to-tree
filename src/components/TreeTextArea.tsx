import * as React from 'react';
import { BinTreeNode } from '../models/TreeNode';
import './TreeInput.scss';

export interface TreeInputProps {
  onChange: (newTreeNode: BinTreeNode) => void;
}
interface TreeInputState {
  treeText: string;
}

export class TreeTextArea extends React.Component<
  TreeInputProps,
  TreeInputState
> {
  constructor(props: TreeInputProps) {
    super(props);
    this.state = {
      treeText: '',
    };
  }

  /**
   * Converts array format binary tree notation to BinTreeNode data structure
   * @param arrayFormat [id, leftChild, rightChild] for example [1, [2], [3, null, [5]]]
   * @returns TreeNode format
   * */
  parseArrayToTree(arrayFormat: any[]): BinTreeNode {
    // recursive solution to convert from array to json tree BinTreeNode
    if (arrayFormat.length === 1) return { id: arrayFormat[0] };

    return {
      id: arrayFormat[0],
      left: arrayFormat[1] ? this.parseArrayToTree(arrayFormat[1]) : null,
      right: arrayFormat[2] ? this.parseArrayToTree(arrayFormat[2]) : null,
    };
  }

  convert = () => {
    let treeArrayFormat: any[] = JSON.parse(this.state.treeText);
    this.props.onChange(this.parseArrayToTree(treeArrayFormat));
  };

  render() {
    return (
      <div>
        <button onClick={this.convert}>Process</button>
        <br />
        <input
          type="text"
          className="array-input"
          onChange={ev => {
            this.setState({
              treeText: ev.target.value,
            });
          }}
        />
      </div>
    );
  }
}
