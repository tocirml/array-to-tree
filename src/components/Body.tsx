import * as React from 'react';
import { IAppState } from '../store/IAppState';
import { observer } from 'mobx-react';
import { TreeInput } from './TreeInput';
import { TreeOutput } from './TreeOutput';
import './Body.scss';
import { useAppStateContext } from '../store/AppState';

interface BodyProps {
  appState: IAppState;
}

const BodyRenderer: React.FunctionComponent<BodyProps> = observer(props => {
  const changeHandler = (newVal: any): void => {
    props.appState.setState({
      ...props.appState,
      treeNode: newVal,
    });
  };

  return (
    <main className="App-body">
      {props.appState!.bodyMessage}
      <TreeInput onChange={changeHandler} />
      <div className="OutputContainer">
        <TreeOutput treeNode={props.appState.treeNode} />
      </div>
    </main>
  );
});

export const Body: React.FunctionComponent<{}> = props => {
  const appState = useAppStateContext();
  return <BodyRenderer appState={appState} />;
};

export default Body;
