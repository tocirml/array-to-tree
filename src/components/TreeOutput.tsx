import * as React from 'react';
import { BinTreeNode } from '../models/TreeNode';

import './TreeOutput.scss';

export interface TreeOutputProps {
  treeNode: BinTreeNode | null | undefined;
  smallestTreeId: string | number;
}

export const TreeOutput: React.FunctionComponent<TreeOutputProps> = props => {
  if (!props.treeNode || !props.treeNode.id) {
    return <div className="treeNode"></div>;
  }
  return (
    <div
      className={`treeNode ${
        // we check what Id matches to paint its border green, problem 3
        props.smallestTreeId === props.treeNode.id && 'smallest-subtree'
      }`}
    >
      <div className="nodeId">{props.treeNode.id}</div>
      {props.treeNode.left || props.treeNode.right ? (
        <div className="nodeChildren">
          <TreeOutput
            treeNode={props.treeNode.left}
            smallestTreeId={props.smallestTreeId}
          />
          <TreeOutput
            treeNode={props.treeNode.right}
            smallestTreeId={props.smallestTreeId}
          />
        </div>
      ) : null}
    </div>
  );
};
