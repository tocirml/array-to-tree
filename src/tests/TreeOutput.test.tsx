import React from 'react';
import { TreeOutput } from '../components/TreeOutput';
import { shallow } from 'enzyme';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import '@testing-library/jest-dom/extend-expect';

const renderTreeOutput = (args?: any) => {
  const defaultProps = {
    onChange: () => {},
  };

  const props = { ...defaultProps, ...args };
  return render(<TreeOutput {...props} />);
};

describe('rendering', () => {
  it('renders without crashing', () => {
    renderTreeOutput();
  });
});

describe('snapshots', () => {
  it('TreeOutput snapshot', () => {
    const treeNode: any = {};
    const smallestTreeId: string = 'root';
    const tree = shallow(
      <TreeOutput treeNode={treeNode} smallestTreeId={smallestTreeId} />
    );
    expect(tree).toMatchSnapshot();
  });
});
