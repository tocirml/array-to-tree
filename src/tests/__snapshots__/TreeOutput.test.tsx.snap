// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots TreeOutput snapshot 1`] = `
<div
  className="treeNode"
/>
`;
